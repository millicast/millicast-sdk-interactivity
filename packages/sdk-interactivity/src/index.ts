// Expose the logger to control when/where to log
export { default as Logger } from 'js-logger';

export * from './types/options';
export * from './types/roomEvents';
export * from './types/sourceType';
export * from './publishedSource';
export * from './publisher';
export * from './room';
export * from './source';
export * from './sourceIdentifier';
