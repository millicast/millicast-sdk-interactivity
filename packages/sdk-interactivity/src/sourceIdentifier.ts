import { SourceType } from './types/sourceType';

/**
 * Represents a source identifier. This class allows to parse the identifier of the source into components like the publisher name, type of the source and name of the source.
 */
export class SourceIdentifier {
    #sourceId: string;

    /** Gets the source identifier. */
    get sourceId(): string {
        return this.#sourceId;
    }

    #publisherName: string;

    /** Gets the name of the publisher. */
    get publisherName(): string {
        return this.#publisherName;
    }

    #sourceType: SourceType;

    /** Gets the type of source published. */
    get sourceType(): SourceType {
        return this.#sourceType;
    }

    #sourceName: string;

    /** Gets the name of the source. */
    get sourceName(): string {
        return this.#sourceName;
    }

    /**
     * Create an instance of the {@link SourceIdentifier} class.
     *
     * @param publisherName Name of the source publisher.
     * @param sourceType Type of source.
     * @param sourceName Name of the source.
     */
    constructor(publisherName: string, sourceType: SourceType, sourceName: string) {
        this.#publisherName = publisherName;
        this.#sourceType = sourceType;
        this.#sourceName = sourceName;

        const obj = {
            pn: publisherName,
            st: sourceType,
            sn: sourceName,
        };
        const jsonStr = JSON.stringify(obj);
        this.#sourceId = btoa(jsonStr);
    }

    /**
     * Get a {@link SourceIdentifier} object extracted from the source ID received from the Dolby.io Real-time Streaming platform.
     *
     * @param sourceId Source Identifier received from the Dolby.io Real-time Streaming platform.
     *
     * @returns The {@link SourceIdentifier} object representing the source identifier.
     */
    public static fromSourceId = (sourceId: string): SourceIdentifier => {
        let publisherName: string;
        let sourceType: SourceType;
        let sourceName: string;

        try {
            const jsonStr = atob(sourceId);
            const { pn, st, sn } = JSON.parse(jsonStr);

            publisherName = pn;
            sourceType = st;
            sourceName = sn;
        } catch (error) {
            // Probably an external source not following the formatting
        }

        const result = new SourceIdentifier(publisherName ?? sourceId, sourceType ?? SourceType.Custom, sourceName ?? sourceId);
        result.#sourceId = sourceId;

        return result;
    };
}
