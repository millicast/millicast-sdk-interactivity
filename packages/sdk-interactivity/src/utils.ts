/**
 * Take a break for a number of seconds.
 * @param seconds Number of seconds to wait.
 */
const sleep = (seconds: number) => new Promise<void>((resolve) => setTimeout(resolve, seconds * 1000));

const isNotDefined = (value: string | null | undefined) => !value || !value.length;

const firstOrUndefined = <T>(array: Array<T>): T => {
    if (array && array.length) {
        return array[0];
    }
};

export { sleep, isNotDefined, firstOrUndefined };
