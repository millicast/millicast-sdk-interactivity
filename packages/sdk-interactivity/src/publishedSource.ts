import { Director, Publish } from '@millicast/sdk';
import { ILogger } from 'js-logger';
import { firstOrUndefined, isNotDefined } from './utils';
import { PublishOptions, PublishWithConstraintsOptions, PublishWithStreamOptions } from './types/options';
import { SourceIdentifier } from './sourceIdentifier';

/**
 * Represents a published source.
 */
export class PublishedSource {
    #logger: ILogger;

    #streamName: string;
    #releaseOnLeave: boolean = true;

    #sourceId: SourceIdentifier;

    /**
     * Gets the {@link SourceIdentifier} object representing the source identifier.
     */
    get sourceId(): SourceIdentifier {
        return this.#sourceId;
    }

    #audioTrack: MediaStreamTrack;

    /**
     * Gets the audio track.
     */
    get audioTrack(): MediaStreamTrack {
        return this.#audioTrack;
    }

    #videoTrack: MediaStreamTrack;

    /**
     * Gets the video track.
     */
    get videoTrack(): MediaStreamTrack {
        return this.#videoTrack;
    }

    #publisher: Publish;

    /**
     * Gets the Millicast {@link !Publish Publish} object used to publish the source to the stream.
     */
    get publisher(): Publish {
        return this.#publisher;
    }

    /** @hidden */
    constructor(logger: ILogger, streamName: string, sourceId: SourceIdentifier) {
        this.#logger = logger;
        this.#streamName = streamName;
        this.#sourceId = sourceId;
    }

    /** @hidden */
    publish = async (options: PublishOptions): Promise<void> => {
        if (isNotDefined(options.publishToken)) throw 'The Publish Token is missing.';

        const tokenGeneratorPublisher = () =>
            Director.getPublisher({
                token: options.publishToken,
                streamName: this.#streamName,
            });

        this.#publisher = new Publish(this.#streamName, tokenGeneratorPublisher, true);

        const optionsWithMediaStream = options as PublishWithStreamOptions;
        if (optionsWithMediaStream.mediaStream) {
            this.#videoTrack = firstOrUndefined(optionsWithMediaStream.mediaStream.getVideoTracks());
            this.#audioTrack = firstOrUndefined(optionsWithMediaStream.mediaStream.getAudioTracks());
            this.#releaseOnLeave = optionsWithMediaStream.releaseOnLeave ?? true;
        } else {
            const optionsWithConstraints = options as PublishWithConstraintsOptions;
            const mediaStream = await navigator.mediaDevices.getUserMedia(optionsWithConstraints.constraints);
            this.#videoTrack = firstOrUndefined(mediaStream.getVideoTracks());
            this.#audioTrack = firstOrUndefined(mediaStream.getAudioTracks());
        }

        const tracks: MediaStreamTrack[] = [];
        if (this.#videoTrack) tracks.push(this.#videoTrack);
        if (this.#audioTrack) tracks.push(this.#audioTrack);

        await this.#publisher.connect({
            mediaStream: tracks,
            sourceId: this.#sourceId.sourceId,
        });
    };

    /** @hidden */
    unpublish = () => {
        if (this.#publisher) {
            this.#logger.info('Stop publishing into the stream');
            this.#publisher.stop();
            this.#publisher = undefined;
        }

        if (this.#releaseOnLeave) {
            this.#logger.info('Stop the audio & video tracks');
            this.#videoTrack?.stop();
            this.#audioTrack?.stop();
        }
    };

    /**
     * Threshold beyond which the publisher is considered to be speaking.
     * @hidden
     */
    readonly IS_SPEAKING_THRESHOLD = 0.002;

    /**
     * Gets if the publisher is considered to be speaking or not.
     *
     * @example
     * ```ts
     * import { Room } from '@millicast/sdk-interactivity';
     *
     * const room = new Room({
     *   streamName,
     *   streamAccountId,
     * });
     *
     * const localSource = await room.connect({
     *   publishToken,
     *   publisherName,
     *   constraints: {
     *     audio: true,
     *     video: true,
     *   },
     * });
     *
     * const videoElement = document.getElementById("local-video");
     * videoElement.srcObject = new MediaStream([localSource.videoTrack]);
     * videoElement.play();
     *
     * setInterval(async () => {
     *   const isSpeaking = await localSource.isSpeaking();
     *   console.log('is speaking', isSpeaking);
     * }, 1000);
     * ```
     *
     * @returns A {@link !Promise Promise} whose fulfillment handler receives a flag indicating if the publisher is considered to be speaking.
     */
    public isSpeaking = async (): Promise<boolean> => {
        if (!this.#audioTrack) return false;

        try {
            const peerConnection: RTCPeerConnection = this.#publisher.getRTCPeerConnection();
            const stats: RTCStatsReport = await peerConnection.getStats(this.#audioTrack);

            const stat = [...stats.values()].find((s) => s.kind === 'audio' && s.audioLevel);
            if (stat) {
                return stat.audioLevel >= this.IS_SPEAKING_THRESHOLD;
            }
        } catch (error) {
            this.#logger.trace('isSpeaking - local', this.#sourceId.sourceId, error);
        }

        return false;
    };

    /**
     * Gets if the audio track is muted or not.
     * If no audio track is available, this returns `true`.
     */
    get isMuted(): boolean {
        if (this.#audioTrack) {
            return !this.#audioTrack.enabled;
        }
        return true;
    }

    /**
     * Mutes the audio track.
     */
    public mute = (): void => {
        if (this.#audioTrack) {
            this.#audioTrack.enabled = false;
        }
    };

    /**
     * Unmutes the audio track.
     */
    public unmute = (): void => {
        if (this.#audioTrack) {
            this.#audioTrack.enabled = true;
        }
    };
}
