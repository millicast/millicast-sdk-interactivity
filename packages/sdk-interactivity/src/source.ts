import { MediaTrackInfo, View } from '@millicast/sdk';
import { ILogger } from 'js-logger';
import { SourceIdentifier } from './sourceIdentifier';
import { firstOrUndefined } from './utils';

export class Source {
    #logger: ILogger;

    #viewer: View;
    #tracks: MediaTrackInfo[];
    #audioMediaId: string;
    #audioTrackId: string;
    #videoMediaId: string;
    #videoTrackId: string;

    /** @ignore */
    get videoMediaId(): string {
        return this.#videoMediaId;
    }

    #identifier: SourceIdentifier;

    /** Gets the source identifier. */
    get identifier(): SourceIdentifier {
        return this.#identifier;
    }

    #mediaStream: MediaStream;

    /** The actual audio / video stream. */
    get mediaStream(): MediaStream {
        return this.#mediaStream;
    }

    /** @hidden */
    constructor(logger: ILogger, viewer: View, sourceId: string, tracks: MediaTrackInfo[]) {
        this.#logger = logger;
        this.#viewer = viewer;
        this.#identifier = SourceIdentifier.fromSourceId(sourceId);
        this.#tracks = tracks;
        this.#mediaStream = new MediaStream();
    }

    /**
     * Request to start receiving the source.
     */
    public receive = async (): Promise<void> => {
        const trackAudio = this.#tracks?.find(({ media }) => media === 'audio');
        if (trackAudio) {
            const audioTransceiver = await this.#viewer.addRemoteTrack('audio', [this.#mediaStream]);
            this.#audioMediaId = audioTransceiver?.mid;
            this.#audioTrackId = trackAudio.trackId;
        }

        const trackVideo = this.#tracks?.find(({ media }) => media === 'video');
        if (trackVideo) {
            const videoTransceiver = await this.#viewer.addRemoteTrack('video', [this.#mediaStream]);
            this.#videoMediaId = videoTransceiver?.mid;
            this.#videoTrackId = trackVideo.trackId;
        }

        await this.project();
    };

    /**
     * Stop receiving the source.
     */
    public stop = async (): Promise<void> => {
        this.#logger.info('About to unproject Source ID:', this.#audioMediaId, this.#videoMediaId);
        await this.#viewer.unproject([this.#audioMediaId, this.#videoMediaId]);
    };

    /**
     * Project a source.
     */
    private project = async (): Promise<void> => {
        const mapping = [];
        if (this.#audioMediaId) {
            mapping.push({
                media: 'audio',
                trackId: this.#audioTrackId,
                mediaId: this.#audioMediaId,
            });
        }
        if (this.#videoMediaId) {
            mapping.push({
                media: 'video',
                trackId: this.#videoTrackId,
                mediaId: this.#videoMediaId,
            });
        }

        this.#logger.info('About to project Source ID:', this.#identifier.sourceId, '-> Track ID:', this.#audioTrackId, this.#videoTrackId);
        await this.#viewer.project(this.#identifier.sourceId, mapping);
    };

    /**
     * Threshold beyond which the publisher is considered to be speaking.
     * @hidden
     */
    readonly IS_SPEAKING_THRESHOLD = 0.002;

    /**
     * Gets if the remote source is considered to be speaking or not.
     *
     * @example
     * ```ts
     * import { Publisher, Room, Source } from '@millicast/sdk-interactivity';
     *
     * const room = new Room({
     *   streamName,
     *   streamAccountId,
     * });
     *
     * room.on('sourceAdded', async (publisher: Publisher, source: Source) => {
     *   // Request to receive the source
     *   await source.receive();
     *
     *   // Display the source on the UI
     *   const videoElement = document.getElementById('remote-video');
     *   videoElement.srcObject = new MediaStream([source.videoTrack]);
     *   videoElement.play();
     *
     *   setInterval(async () => {
     *     const isSpeaking = await source.isSpeaking();
     *     console.log('is speaking', isSpeaking);
     *   }, 1000);
     * });
     *
     * await room.watch();
     * ```
     *
     * @returns A {@link !Promise Promise} whose fulfillment handler receives a flag indicating if the publisher is considered to be speaking.
     */
    public isSpeaking = async (): Promise<boolean> => {
        if (!this.#audioMediaId) return false;

        try {
            const peerConnection: RTCPeerConnection = this.#viewer.getRTCPeerConnection();
            const stats: RTCStatsReport = await peerConnection.getStats();

            const stat = [...stats.values()].find((s) => s.kind === 'audio' && s.mid === this.#audioMediaId && s.audioLevel);
            if (stat) {
                return stat.audioLevel >= this.IS_SPEAKING_THRESHOLD;
            }
        } catch (error) {
            this.#logger.trace('isSpeaking - remote', this.#identifier.sourceId, error);
        }

        return false;
    };

    /**
     * Gets if the audio track is muted or not.
     * If no audio track is available, this returns `true`.
     */
    get isMuted(): boolean {
        const audioTrack = firstOrUndefined(this.#mediaStream.getAudioTracks());
        if (audioTrack) {
            return !audioTrack.enabled;
        }
        return true;
    }

    /**
     * Mutes the audio track.
     */
    public mute = (): void => {
        const audioTrack = firstOrUndefined(this.#mediaStream.getAudioTracks());
        if (audioTrack) {
            audioTrack.enabled = false;
        }
    };

    /**
     * Unmutes the audio track.
     */
    public unmute = (): void => {
        const audioTrack = firstOrUndefined(this.#mediaStream.getAudioTracks());
        if (audioTrack) {
            audioTrack.enabled = true;
        }
    };
}
