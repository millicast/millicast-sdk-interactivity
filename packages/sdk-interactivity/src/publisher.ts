import { Source } from './source';

export class Publisher {
    #name: string;

    /** Gets the name of the publisher. */
    get name(): string {
        return this.#name;
    }

    #sources: Map<string, Source>;

    /** Gets the list of available sources for this publisher. */
    get sources(): Source[] {
        return [...this.#sources.values()];
    }

    /** @hidden */
    constructor(name: string) {
        this.#name = name;
        this.#sources = new Map<string, Source>();
    }

    /** @hidden */
    addSource = (source: Source): void => {
        this.#sources.set(source.identifier.sourceId, source);
    };

    /** @hidden */
    removeSource = async (sourceId: string): Promise<void> => {
        if (this.#sources.has(sourceId)) {
            const source = this.#sources.get(sourceId);
            await source.stop();

            this.#sources.delete(sourceId);
        }
    };
}
