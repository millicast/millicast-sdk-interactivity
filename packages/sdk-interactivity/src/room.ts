import { Director, View, BroadcastEvent, MediaStreamSource, ViewerCount } from '@millicast/sdk';
import { EventEmitter } from 'events';
import Logger, { ILogger } from 'js-logger';
import { isNotDefined } from './utils';
import { ConnectOptions, PublishOptions, StreamInformation, WatchOptions } from './types/options';
import { RoomEvents } from './types/roomEvents';
import { Publisher } from './publisher';
import { PublishedSource } from './publishedSource';
import { Source } from './source';
import { SourceIdentifier } from './sourceIdentifier';

export interface Room {
    /**
     * Adds the `listener` function to the end of the listeners array for the
     * event named `eventName`. No checks are made to see if the `listener` has
     * already been added. Multiple calls passing the same combination of `eventName`and `listener` will result in the `listener` being added, and called, multiple
     * times.
     *
     * Returns a reference to the `EventEmitter`, so that calls can be chained.
     * @param eventName The name of the event.
     * @param listener The callback function.
     */
    on<N extends keyof RoomEvents>(eventName: N, listener: RoomEvents[N]): this;

    /** @hidden */
    emit<N extends keyof RoomEvents>(eventName: N, ...args: Parameters<RoomEvents[N]>): boolean;
}

/**
 * Representation of a stream. This class is the starting point for the SDK.
 * Create a instance of this class, connect to the Dolby Millicast platform and start receiving & publishing streams.
 *
 * @example
 * ```ts
 * import { Room } from '@millicast/sdk-interactivity';
 *
 * const room = new Room({
 *   streamName,
 *   streamAccountId,
 * });
 *
 * // Start publishing to the stream and listen to the available sources
 * await room.connect({
 *   publisherName,
 *   publishToken,
 *   constraints: {
 *     audio: true,
 *     video: true,
 *   },
 * });
 * ```
 */
export class Room extends EventEmitter implements Room {
    #logger: ILogger;
    #streamInfo: StreamInformation;
    #publisherName: string;

    #viewer: View;

    /**
     * Gets the Millicast {@link !View View} object once the SDK has been initialized.
     */
    get viewer(): View {
        return this.#viewer;
    }

    #publishers: Map<string, Publisher> = new Map<string, Publisher>();

    /**
     * Gets the list of {@link Publisher} objects that are currently publishing a stream.
     */
    get publishers(): Array<Publisher> {
        return [...this.#publishers.values()];
    }

    #publishedSources: Map<string, PublishedSource> = new Map<string, PublishedSource>();

    /**
     * Gets the list of {@link PublishedSource} objects that are currently published to the stream from the local device.
     */
    get publishedSources(): Array<PublishedSource> {
        return [...this.#publishedSources.values()];
    }

    /** Maximum number of sources that can be published from a single client. */
    readonly MAX_SOURCES: number = 5;

    /**
     * Create an instance of the {@link Room} class.
     *
     * @param streamInfo Information about the stream to connect to.
     */
    constructor(streamInfo: StreamInformation) {
        if (isNotDefined(streamInfo.streamName)) throw 'The Stream Name is missing.';
        if (isNotDefined(streamInfo.streamAccountId)) throw 'The Stream Account ID is missing.';

        super();

        this.#logger = Logger.get('Interactivity');
        this.#streamInfo = streamInfo;
    }

    private onBroadcastEvent = async (event: BroadcastEvent) => {
        this.#logger.debug('onBroadcastEvent', event);

        if (event.name === 'active') {
            const { sourceId, tracks } = event.data as MediaStreamSource;
            this.#logger.debug(`New source available - ${sourceId}`);
            if (this.#publishedSources.has(sourceId)) {
                this.#logger.debug('This is a local source so ignoring it.');
                return;
            }

            const source = new Source(this.#logger, this.#viewer, sourceId, tracks);

            let publisher: Publisher;
            if (this.#publishers.has(source.identifier.publisherName)) {
                publisher = this.#publishers.get(source.identifier.publisherName);
            } else {
                publisher = new Publisher(source.identifier.publisherName);
                this.#publishers.set(source.identifier.publisherName, publisher);
                this.emit('publisherJoined', publisher);
            }

            publisher.addSource(source);
            this.emit('sourceAdded', publisher, source);
        } else if (event.name === 'inactive') {
            const { sourceId } = event.data as MediaStreamSource;
            const sourceIdentifier = SourceIdentifier.fromSourceId(sourceId);

            this.#logger.debug(`The source has been stopped - ${sourceId}`);
            if (sourceIdentifier.publisherName === this.#publisherName) {
                this.#logger.debug('This is a local source so ignoring it.');
                return;
            }

            if (this.#publishers.has(sourceIdentifier.publisherName)) {
                const publisher = this.#publishers.get(sourceIdentifier.publisherName);
                publisher.removeSource(sourceId);

                this.emit('sourceRemoved', publisher, sourceIdentifier);

                if (publisher.sources.length <= 0) {
                    this.#publishers.delete(sourceIdentifier.publisherName);
                    this.emit('publisherLeft', publisher);
                }
            }
        } else if (event.name === 'viewercount') {
            const { viewercount } = event.data as ViewerCount;
            this.emit('viewercount', viewercount);
        }
    };

    /**
     * Connects to the stream. This function will publish a stream and listen to the sources.
     *
     * @param options Options to connect to the stream.
     *
     * @returns A {@link !Promise Promise} whose fulfillment handler receives a {@link PublishedSource} object when the source has successfully been published to the platform.
     */
    public connect = async (options: ConnectOptions): Promise<PublishedSource> => {
        this.#publisherName = options.publisherName;

        // Publish the source to the platform
        const publishedSource: PublishedSource = await this.publish(options);

        // Start watching the other sources
        await this.watch(options);

        return publishedSource;
    };

    /**
     * Start watching the stream.
     *
     * @param options Additional options to watch to the stream.
     */
    public watch = async (options?: WatchOptions): Promise<void> => {
        const tokenGeneratorViewer = () =>
            Director.getSubscriber({
                streamName: this.#streamInfo.streamName,
                streamAccountId: this.#streamInfo.streamAccountId,
                subscriberToken: options?.subscriberToken,
            });

        this.#viewer = new View(undefined, tokenGeneratorViewer, null, true);

        this.#viewer.on('broadcastEvent', this.onBroadcastEvent);

        await this.#viewer.connect({
            events: ['active', 'inactive', 'viewercount', 'layers'],
        });
    };

    /**
     * Publishes a video / audio feed to a stream.
     *
     * @param options Information about the stream to publish to.
     *
     * @returns A {@link !Promise Promise} whose fulfillment handler receives a {@link PublishedSource} object when the source has successfully been published to the platform.
     *
     * @throws {@link !Error Error} if you have not set the `publisherName` when connecting to the stream.
     * @throws {@link !Error Error} if you've reached the maximum number of sources that can be published from a single publisher.
     */
    public publish = async (options: PublishOptions): Promise<PublishedSource> => {
        if (isNotDefined(this.#publisherName)) {
            throw Error('The name of the publisher is missing. Reconnect to the stream.');
        }

        if (this.#publishedSources.size >= this.MAX_SOURCES) {
            throw Error(`Maximum number of ${this.MAX_SOURCES} sources reached!`);
        }

        const sourceId = new SourceIdentifier(this.#publisherName, options.sourceType, options.sourceName ?? options.sourceType);

        const publishedSource = new PublishedSource(this.#logger, this.#streamInfo.streamName, sourceId);
        await publishedSource.publish(options);

        this.#publishedSources.set(sourceId.sourceId, publishedSource);

        return publishedSource;
    };

    /**
     * Stops publishing the source to the stream.
     *
     * @param publishedSource Source that's currently being published.
     */
    public unpublish = (publishedSource: PublishedSource) => {
        publishedSource.unpublish();
        this.#publishedSources.delete(publishedSource.sourceId.sourceId);
    };

    /**
     * Leaves the stream. Will stop publishing into the stream and will stop listening to it.
     *
     * _Note_, the audio / video stream will be stopped to release allocated resources.
     *
     * @example
     * ```js
     * room.leave();
     * ```
     */
    public leave = (): void => {
        this.#publishedSources.forEach((publishedSource) => {
            publishedSource.unpublish();
        });

        if (this.#viewer) {
            this.#logger.info('Stop listening to the stream');
            this.#viewer.stop();
        }

        this.#publishers.clear();
    };
}
