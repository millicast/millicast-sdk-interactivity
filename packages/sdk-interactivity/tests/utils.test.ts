import * as utils from '../src/utils';

describe('utils', () => {
    test('utils should match the snapshot', () => {
        expect(utils).toMatchSnapshot();
    });

    test('utils - sleep should take a nap', async () => {
        const seconds = 1.234;

        const start = performance.now();
        await utils.sleep(seconds);
        const stop = performance.now();

        expect(stop - start).toBeGreaterThanOrEqual(seconds * 1000 * 0.99);
    });

    test('utils - isNotDefined', async () => {
        expect(utils.isNotDefined('')).toBe(true);
        expect(utils.isNotDefined(null)).toBe(true);
        expect(utils.isNotDefined(undefined)).toBe(true);
        expect(utils.isNotDefined('ok')).toBe(false);
    });

    test('utils - firstOrUndefined', async () => {
        expect(utils.firstOrUndefined(null as unknown as any[])).toBeUndefined();
        expect(utils.firstOrUndefined(undefined as unknown as any[])).toBeUndefined();
        expect(utils.firstOrUndefined([])).toBeUndefined();

        expect(utils.firstOrUndefined(['a'])).toBe('a');
        expect(utils.firstOrUndefined(['a', 'b'])).toBe('a');
    });
});
