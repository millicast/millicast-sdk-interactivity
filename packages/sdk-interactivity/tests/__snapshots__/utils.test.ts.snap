// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils utils should match the snapshot 1`] = `
{
  "firstOrUndefined": [Function],
  "isNotDefined": [Function],
  "sleep": [Function],
}
`;
